### PR TITLE
[Flang2] Support for vector always loop directive

### PR DESCRIPTION
**Pull request by Chirag-Khandelwal**
This PR adds loop directive support for !dir$ vector always (XBIT/XFLAG: 191/0x2).
It uses the llvm.loop.parallel_accesses metadata with the companion - llvm.access.group metadata.

(Created new PR as I had deleted the fork of the older one due to my misunderstanding).

@kiranchandramohan please add the reviewers here.

### CODE REVIEW

File: test/directives/vector_directive1.f90 [hunk 1]
```diff
@@ -1,4 +1,4 @@
-! RUN: %flang -c %s 2>&1 | FileCheck %s --check-prefix=CHECK-NO-CLAUSE
+! RUN: %flang -S -emit-llvm %s 2>&1 -o - | FileCheck %s --check-prefix=CHECK-NO-CLAUSE
```
Comment (kiranchandramohan) [inline test/directives/vector_directive1.f90:1]: Is this change necessary?

File: test/directives/vector_directive3.f90 [hunk 2]
```diff
@@ -1,15 +1,22 @@
...
-    arr3(i) = arr1(i) - arr2(i)
-  end do
-end subroutine
-! CHECK-NOT: F90-S-0602
-! CHECK-NOT: F90-S-0603
+  !DIR$ VECTOR ALWAYS
+  DO i=1,ub
+      myarr1(i) = myarr1(i)+myarr2(i)
+  END DO
+END SUBROUTINE
```
Comment (kiranchandramohan) [inline test/directives/vector_directive3.f90:22]: Nit: no newline.

File: test/directives/vector_directive3.f90 [hunk 3]
```diff
@@ -1,15 +1,22 @@
-! RUN: %flang -c %s 2>&1 | FileCheck %s -allow-empty --check-prefix=CHECK
+!! check for pragma support for !dir$ vector always
+!RUN: %flang -S -O2 -emit-llvm %s -o - | FileCheck %s
+!CHECK: define void @sumsimd_{{.*$}}
+!CHECK: {{.*}}!llvm.access.group{{.*}}
+!CHECK: vector.ph:{{.*}}
+!CHECK: vector.body:{{.*}}
+!CHECK: {{.*}}shufflevector{{.*}}
+!CHECK: {{.*}}add <2 x i64>{{.*}}
+!CHECK: {{.*}}"llvm.loop.parallel_accesses"{{.*}}
```
Comment (kiranchandramohan) [inline test/directives/vector_directive3.f90:9]: Should check that the access mentioned in the parallel_accesses matches the access.group above.

File: tools/shared/pragma.c [hunk 4]
```diff
@@ -655,16 +655,14 @@ do_sw(void)
...
-          gbl.nowarn = backup_nowarn;
+          bclr(DIR_OFFSET(currdir, x[19]), 0x18);
+          bset(DIR_OFFSET(currdir, x[19]), 0x400);
           break;
         }
         LCASE(ctok);
         if (strcmp(ctok, "always") == 0) {
           bclr(DIR_OFFSET(currdir, x[19]), 0x18);
-          bset(DIR_OFFSET(currdir, x[19]), 0x400);
+          bset(DIR_OFFSET(currdir, x[191]), 0x2);
```
Comment (kiranchandramohan) [inline tools/shared/pragma.c:665]: Why was a change required?

**Review by kiranchandramohan — COMMENTED** [review]
A few quick comments.

File: tools/flang2/flang2exe/cgmain.cpp [hunk 5]
```diff
@@ -949,6 +968,20 @@ assign_fortran_storage_classes(void)
...
+ * when vector always pragma is specified, "llvm.loop.parallel_accesses" metadata has
+ * to be generated along with "llvm.access.group" for each load/store instructions.
+ */
+INLINE static LL_MDRef
+cons_loop_parallel_accesses_metadata(void)
+{
+  LL_MDRef lvcomp[2];
+
+  lvcomp[0] = ll_get_md_string(cpu_llvm_module, "llvm.loop.parallel_accesses");
+  lvcomp[1] = cached_access_group_metadata;
```
Comment (kiranchandramohan) [inline tools/flang2/flang2exe/cgmain.cpp:981]: Is it OK to reuse the access-group-metadata always?

File: tools/flang2/flang2exe/cgmain.cpp [hunk 6]
```diff
@@ -1344,6 +1392,36 @@ remove_dead_instrs(void)
   }
 }
 
+/*
+ * Check if the break instruction is having a loop pragma
```
Comment (kiranchandramohan) [inline tools/flang2/flang2exe/cgmain.cpp:1396]: break or the branch instruction (back edge)?

File: test/directives/vector_directive1.f90 (hunk 1 above)
Comment (Chirag-Khandelwal) [reply test/directives/vector_directive1.f90:1]: The `-o -` part is necessary. Without that, the test fails with the following:
```
FileCheck error: '<stdin>' is empty.
FileCheck command line:  /home/chirag/repos/chirag-flang/build/bin/FileCheck /home/chirag/repos/chirag-flang/test/directives/vector_directive1.f90 --check-prefix=CHECK-NO-CLAUSE
```
I'll remove the `-S -emit-llvm` part.

File: test/directives/vector_directive3.f90 (hunk 3 above)
Comment (Chirag-Khandelwal) [reply test/directives/vector_directive3.f90]: Done.

File: test/directives/vector_directive3.f90 (hunk 2 above)
Comment (Chirag-Khandelwal) [reply test/directives/vector_directive3.f90]: Done.

File: test/directives/vector_directive1.f90 (hunk 1 above)
Comment (Chirag-Khandelwal) [reply test/directives/vector_directive1.f90]: Done.

File: tools/flang2/flang2exe/cgmain.cpp (hunk 6 above)
Comment (Chirag-Khandelwal) [reply tools/flang2/flang2exe/cgmain.cpp]: Renamed to `branch`.

File: tools/flang2/flang2exe/cgmain.cpp (hunk 5 above)
Comment (Chirag-Khandelwal) [reply tools/flang2/flang2exe/cgmain.cpp:981]: Yes it should be fine. Even if for example we have multiple functions with different loops using this pragma, the metadata is actually replicated for each set of the `llvm.access.group` (therefore multiple instances of `llvm.loop.parallel_accesses` metadata is generated).

File: tools/shared/pragma.c (hunk 4 above)
Comment (Chirag-Khandelwal) [reply tools/shared/pragma.c:665]: Not required, per say, but I did that so as to not alter any existing work.
If desired, I shall use the same xflag/xbit values.

File: test/directives/vector_directive1.f90 [hunk 7]
```diff
@@ -11,4 +11,5 @@ subroutine add(arr1,arr2,arr3,N)
     arr3(i) = arr1(i) - arr2(i)
   end do
 end subroutine
-! CHECK-NO-CLAUSE: F90-W-0602-No clause specified for the vector directive. Note: Only the always clause is supported.
+! CHECK-NO-CLAUSE-NOT: F90-S-0602
+! CHECK-NO-CLAUSE-NOT: F90-S-0603.
```
Comment (bryanpkc) [inline test/directives/vector_directive1.f90:15]: Is the period at the end intentional?

Does this change mean that Flang now supports `!dir$ vector` (without the `always` clause)? If that's the case, you might also want to add `-emit-llvm` and check the IR to verify that vectorization happens.

File: tools/shared/pragma.c (hunk 4 above)
Comment (bryanpkc) [reply tools/shared/pragma.c:665]: This block is not executed if `craydir` is false. I don't think it gets set to true by `!dir$ vector`? Why isn't the same change needed on line 674?

Comment (Chirag-Khandelwal) [reply tools/shared/pragma.c:665]: The change was not required on line 674 because `191, 0x2` is exclusively for vector always.
Everywhere else, `19, 0x400` is used since that is the one for `!dir$ vector`.
Since line 674 deals with `!dir$ vector` alone, `191, 0x2` is not used there.

Unless, you want me to call `bclr()` on that bit at/near line `674`?

File: test/directives/vector_directive1.f90 (hunk 7 above)
Comment (Chirag-Khandelwal) [reply test/directives/vector_directive1.f90]: Yes `!dir$ vector` is now functional.
Added the llvm metadata + load/store test.
Fixed the period at the end of line 15.

**Comment by Chirag-Khandelwal** [comment]
Updated the xbit to `191, 0x4` as `0x2` was taken by forceinline pragma. (#1017)

**Comment by kiranchandramohan** [comment]
ping @bryanpkc @kiranchandramohan

File: test/directives/vector_directive1.f90 [hunk 8]
```diff
@@ -11,4 +13,9 @@ subroutine add(arr1,arr2,arr3,N)
     arr3(i) = arr1(i) - arr2(i)
   end do
 end subroutine
-! CHECK-NO-CLAUSE: F90-W-0602-No clause specified for the vector directive. Note: Only the always clause is supported.
+! CHECK-NO-CLAUSE-NOT: F90-S-0602
+! CHECK-NO-CLAUSE-NOT: F90-S-0603
+
+! METADATA: !"llvm.loop.vectorize.enable", i1 true
+! CHECK: load <[[VF:[0-9]+]] x i32>
+! CHECK: store <[[VF]] x i32>
```
Comment (bryanpkc) [inline test/directives/vector_directive1.f90:21]: Missing newline at end of file.

**Review by michalpasztamobica — APPROVED** [review]
I cherry-picked the most recent status to [my branch](https://github.com/michalpasztamobica/flang/tree/loop-line-fix-parallel_loop_access) and still got the tests passing alright, so green light from my side. @Chirag-Khandelwal , if you have time to look at my branch and give your feedback - I will be very grateful.

**Review by bryanpkc — APPROVED** [review]
Sorry for taking so long, but finally I understand the logic. I'll go ahead and merge the branch.
